internal/oci/notation.go

```diff
@@ -0,0 +1,241 @@
+	opts     []remote.Option
+	insecure bool
+}
+
+type trustStore struct {
+	cert []byte
+}
+
+// GetCertificates implements truststore.X509TrustStore.
+func (s trustStore) GetCertificates(ctx context.Context, storeType truststore.Type, namedStore string) ([]*x509.Certificate, error) {
+	raw := s.cert
+	block, _ := pem.Decode(raw)
+	if block != nil {
+		raw = block.Bytes
+	}
+
+	certs, err := x509.ParseCertificates(raw)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse certificate '%s': %s", namedStore, err)
+	}
+
+	return certs, nil
+}
+
+// NewNotaryVerifier initializes a new NotaryVerifier
+func NewNotaryVerifier(opts ...Options) (*NotaryVerifier, error) {
+	o := options{}
+	for _, opt := range opts {
+		opt(&o)
+	}
+
+	store := &trustStore{
+		cert: o.PublicKey,
+	}
+
+	verifier, err := verifier.New(o.TrustStore, store, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	switch {
+	case o.Auth != nil && o.Keychain != nil:
+		return nil, fmt.Errorf("please provide either authn.Authenticator or authn.Keychain")
+	case o.Auth == nil:
+		o.Auth = authn.Anonymous
+	}
+
+	return &NotaryVerifier{
+		auth:     o.Auth,
+		keychain: o.Keychain,
+		verifier: &verifier,
+		opts:     o.ROpt,
+		insecure: o.Insecure,
+	}, nil
+}
+
+// Verify verifies the authenticity of the given ref OCI image.
+// It returns a boolean indicating if the verification was successful.
+// It returns an error if the verification fails, nil otherwise.
+func (v *NotaryVerifier) Verify(ctx context.Context, ref name.Reference) (bool, error) {
+	url := ref.Name()
+	remoteRepo, err := oras.NewRepository(url)
+	if err != nil {
+		return false, err
+	}
+	remoteRepo.PlainHTTP = v.insecure
+
+	repo := registry.NewRepository(remoteRepo)
+
+	credentialProvider := func(ctx context.Context, registry string) (oauth.Credential, error) {
+		return oauth.EmptyCredential, nil
+	}
+
+	auth := v.auth
+	if v.keychain != nil {
+		source := stringResource{url}
+
+		auth, err = v.keychain.Resolve(source)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	if auth != authn.Anonymous {
+		authConfig, err := auth.Authorization()
+		if err != nil {
+			return false, err
+		}
+
+		credentialProvider = func(ctx context.Context, registry string) (oauth.Credential, error) {
+			if authConfig.Username != "" || authConfig.Password != "" || authConfig.IdentityToken != "" {
+				return oauth.Credential{
+					Username:     authConfig.Username,
+					Password:     authConfig.Password,
+					RefreshToken: authConfig.IdentityToken,
+				}, nil
+			}
+			return oauth.EmptyCredential, nil
+		}
+	}
+
+	repoClient := &oauth.Client{
+		Header: http.Header{
+			"User-Agent": {"flux"},
+		},
+		Cache:      oauth.NewCache(),
+		Credential: credentialProvider,
+	}
+
+	remoteRepo.Client = repoClient
+
+	i, err := remote.Image(ref, v.opts...)
+	if err != nil {
+		return false, err
+	}
+
+	d, err := i.Digest()
+	if err != nil {
+		return false, err
+	}
+
+	repoUrl := ""
+
+	lastIndex := strings.LastIndex(url, ":")
```

Comment:
when `strings.LastIndex` returns `-1`, the next line will panic because go slice indexes cannot be negative. Even if having `-1` might mean an invalid reference, we should guard against it. Or maybe use `strings.Cut()`

Reply:
I found a bug with my approach. As it is, if the user sets `ref.digest` in the `OCIRepository` configuration, this will always fail further down and never retrieve the signature. I assumed the user will always use tags which is not the case. I have a fix (with unit tests) so this will change slightly. I think it's good to check for -1 in the unlikely chance it may cause a panic just to be safe.

Reply:
Resolved in [0c31364](https://github.com/fluxcd/source-controller/pull/1075/commits/0c313644720b9cbcb2d7889213f222452bd943a8).